[Joel Chan Zhi Yang] iP

JoelChanZhiYang opened the pull request:
# DukePro

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's 
* text-based
* easy to learn
* ~~FAST~~ *SUPER FAST* to use

All you need to do is,
1. download it from [here](www.google.com).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😀 

And it is **FREE**!

Features: 

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static voide main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

JoelChanZhiYang committed:
Finished level 1. Added Greet, echo and list

JoelChanZhiYang committed:
Added the function to add to the todo list as well as listing off the todo list.

JoelChanZhiYang committed:
Added a mark as done feature

JoelChanZhiYang committed:
Added todos, events and deadlines

JoelChanZhiYang committed:
Fixed the bug where a new scanner was instantiated after every response. Now only one scanner is used. Added semi-automated testing. Could still be improved

JoelChanZhiYang committed:
Split the logic of each command into different functions

JoelChanZhiYang committed:
Removed commented out code

JoelChanZhiYang committed:
Added error handling

JoelChanZhiYang committed:
Added testing for level 5

JoelChanZhiYang committed:
Added the delete function. Added tests for delete function as well

JoelChanZhiYang committed:
Add persistance to tasks

JoelChanZhiYang committed:
Modify text-ui-test to align with new outputs

JoelChanZhiYang committed:
Add date-time parsing ability for Event and Deadline class.

JoelChanZhiYang committed:
Merge branch 'branch-Level-7'

JoelChanZhiYang committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Task.java

JoelChanZhiYang committed:
Refactor code into classes

All the logic were in the Duke class.

Extracting the common behavior out to classes improves readability

Let's create classes to manage the UI, parsing, command logic and storage logic

JoelChanZhiYang committed:
Organise classes into packages

Classes were all in one 'Java' folder.

Packages will make the source directory neater and easier to maintain.

Let's organise the classes into 4 packages. Util, task, exception and command

JoelChanZhiYang committed:
Add some JUnit tests for TodoCommand, Todo and Parser

JoelChanZhiYang committed:
Add JavaDoc comments

JoelChanZhiYang committed:
Tweaked ByeCommand to fit coding standard

JoelChanZhiYang committed:
Tweak the code to comply with a coding standard

JoelChanZhiYang committed:
Add the ability for user to search for items in their tasks

JoelChanZhiYang committed:
Merge branch 'branch-A-JavaDoc'

JoelChanZhiYang committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DoneCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/exception/DukeException.java
#	src/main/java/duke/exception/NoActionException.java
#	src/main/java/duke/exception/NoTimeException.java
#	src/main/java/duke/exception/TaskCompletedException.java
#	src/main/java/duke/exception/TaskNotFoundException.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/util/Parser.java
#	src/main/java/duke/util/Storage.java
#	src/main/java/duke/util/TaskList.java
#	src/main/java/duke/util/Ui.java

JoelChanZhiYang committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/util/Parser.java
#	src/main/java/duke/util/Ui.java

JoelChanZhiYang committed:
Import FindCommand to Parser

JoelChanZhiYang committed:
Merge branch 'add-gradle-support'

JoelChanZhiYang committed:
Change mainClassNme in build.gradle

JoelChanZhiYang committed:
Add checkstyle to Gradle and fixed any coding style violations

JoelChanZhiYang committed:
Merge branch 'branch-A-CheckStyle'

JoelChanZhiYang committed:
Add a GUI to Duke

JoelChanZhiYang committed:
Add documentation and tweaked code to adhere to styling guide

JoelChanZhiYang committed:
Merge branch 'branch-Level-10'

nniiggeell commented on src/main/java/duke/command/TodoCommand.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    @Override
+    public CommandResult execute(TaskList tasks, Storage storage)
+            throws NoActionException, SaveFileException {
```
Can throws NoActionException be on the same line as method declaration? It might be clearer.

nniiggeell commented on src/main/java/duke/ui/MainWindow.java:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @param result CommandResult after executing the command.
+     */
+    private void handleCommandResult(CommandResult result) {
```
This is very clear!

nniiggeell reviewed (commented):
Your code looks very good! It's very clear and very minimal code quality issues

jeromewjj commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,76 @@
+            throw new NoTimeException(
+                    "Command 'deadline' requires a deadline to be specified. Use /by to specify a deadline.");
+        }
+        LocalDateTime deadline = Parser.parseDate(deadlineInputs[1].trim());
```
I like that you provide detailed error messages base on the error created so that user can know what is the error.

jeromewjj commented on src/main/java/duke/util/Reply.java:
```diff
@@ -0,0 +1,110 @@
+    private final Scanner scanner;
+
+    /**
+     * Constructor of Ui class.
```
I think you forgot to delete this.

jeromewjj commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,86 @@
+    public static Command parse(String[] fullCommand) throws InvalidInputException {
+        String command = fullCommand[0];
+        String action = fullCommand[1];
+        switch (command) {
+        case "todo":
+            return new TodoCommand(action);
+        case "deadline":
+            return new DeadlineCommand(action);
+        case "event":
+            return new EventCommand(action);
+        case "list":
+            return new ListCommand();
+        case "done":
+            return new DoneCommand(action);
+        case "bye":
+            return new ByeCommand();
+        case "delete":
+            return new DeleteCommand(action);
+        case "find":
+            return new FindCommand(action);
```
Should there be break statement?

jeromewjj commented on src/main/java/duke/util/Reply.java:
```diff
@@ -0,0 +1,110 @@
+
+    /**
+     * Shows a goodbye message.
+     */
```
I think you forgot to add @return in your JavaDocs

jeromewjj reviewed (commented):
Overall, your code looks very clean and doesn't seem to violate the coding standard. 😃

JoelChanZhiYang committed:
Add assertions to verify any assumptions in the method

Several methods make assumptions about the input without checking if they are true

Assertions could catch bugs that could be caused by these assumptions.

Assertion provide a more elegant solution to check these assumptions as compared to if-else statements.

Let's add assertions to verify the assumptions in all methods.

JoelChanZhiYang committed:
Refactor code class Storage to structure clearer

The parsing of the save file was done inside storage class with a lot of code duplication.

An method to do the parsing would make the code more readable.

Adding a new method in the Parser class called parseSaveFileLine would abstract out this logic from the Storage class.

Let's move the logic from Storage into this new method in the Parser class

JoelChanZhiYang committed:
Merge pull request #5 from JoelChanZhiYang/branch-A-Assertions

Add assertions to verify any assumptions in the method

JoelChanZhiYang committed:
Merge remote-tracking branch 'origin/master' into branch-A-CodeQuality

JoelChanZhiYang committed:
Merge pull request #4 from JoelChanZhiYang/branch-A-CodeQuality

Refactor code class Storage to structure clearer

JoelChanZhiYang committed:
Add GitHub Actions yml configuration

JoelChanZhiYang committed:
Change name of directory to set configuration

JoelChanZhiYang committed:
Add unit tests to Github Actions CI

JoelChanZhiYang committed:
Change the title of the gradle tests

JoelChanZhiYang committed:
Merge pull request #6 from JoelChanZhiYang/branch-A-CI

Add GitHub Actions yml configuration

JoelChanZhiYang committed:
Add tagging functionality to Duke

JoelChanZhiYang committed:
Modify tests to work with new additions

JoelChanZhiYang committed:
Merge branch 'branch-C-Tagging'

JoelChanZhiYang committed:
Improve the padding and the margin of the GUI

JoelChanZhiYang committed:
Merge branch 'branch-A-BetterGui'

JoelChanZhiYang committed:
Update README.md

JoelChanZhiYang committed:
Add Ui image

JoelChanZhiYang committed:
Set theme jekyll-theme-slate

JoelChanZhiYang committed:
Add a welcome message to Duke

JoelChanZhiYang committed:
Merge branch 'branch-WelcomeMessage'
